core/src/replay_stage.rs

```diff
@@ -850,34 +850,50 @@ impl ReplayStage {
-            if !progress
-                .get_propagated_stats(latest_leader_slot)
-                .map(|stats| stats.is_propagated)
-                .unwrap_or(true)
-            {
-                warn!("Slot not propagated: slot={}", latest_leader_slot);
-                let retransmit_info = progress.get_retransmit_info(latest_leader_slot).unwrap();
-                if retransmit_info.reached_retransmit_threshold() {
-                    info!(
-                        "Retrying retransmit: start_slot={} latest_leader_slot={} retransmit_info={:?}",
-                        start_slot, latest_leader_slot, &retransmit_info,
-                    );
-                    datapoint_info!(
-                        "replay_stage-retransmit-timing-based",
-                        ("slot", latest_leader_slot, i64),
-                        ("retry_iteration", retransmit_info.retry_iteration, i64),
-                    );
-                    let _ = retransmit_slots_sender.send(latest_leader_slot);
-                    retransmit_info.increment_retry_iteration();
-                } else {
-                    info!("Bypass retry: {:?}", &retransmit_info);
+        if let (false, Some(latest_leader_slot)) =
+            progress.get_leader_propagation_slot_must_exist(start_slot)
+        {
+            info!(
+                "Slot not propagated: start_slot={} latest_leader_slot={}",
+                start_slot, latest_leader_slot
+            );
+
+            let first_leader_group_slot =
+                Self::first_of_consecutive_leader_slots(latest_leader_slot);
+
+            for slot in first_leader_group_slot..=latest_leader_slot {
+                if !progress.is_propagated(slot).unwrap_or(true) {
+                    if let Some(retransmit_info) = progress.get_retransmit_info(slot) {
+                        if retransmit_info.reached_retransmit_threshold() {
+                            info!(
+                                "Retrying retransmit: start_slot={} latest_leader_slot={} slot={} retransmit_info={:?}",
+                                start_slot,
+                                latest_leader_slot,
+                                slot,
+                                &retransmit_info,
+                            );
+                            datapoint_info!(
+                                "replay_stage-retransmit-timing-based",
+                                ("latest_leader_slot", latest_leader_slot, i64),
+                                ("slot", slot, i64),
+                                ("retry_iteration", retransmit_info.retry_iteration, i64),
+                            );
+                            let _ = retransmit_slots_sender.send(slot);
+                            retransmit_info.increment_retry_iteration();
+                        } else {
+                            info!("Bypass retry: {:?}", &retransmit_info);
```

Comment:
This might be a little spammy to log every time we miss the retransmit threshold, because this replay loop runs may run every few milliseconds, but the retransmit thresholds is on the order of seconds

Reply:
changed to `debug!`